hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerRegistryImpl.java

```diff
@@ -160,9 +160,52 @@ public void prepare(MetadataImplementor metadata) {
 		}
 	}
 
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	private synchronized <T> EventListenerGroupImpl<T> getOrCreateEventListenerGroup(EventType<T> eventType) {
+		final int sizeOriginal = this.registeredEventListeners.length;
+		if ( eventType.ordinal() < sizeOriginal ) {
+			final EventListenerGroupImpl registeredEventListener = registeredEventListeners[ eventType.ordinal() ];
+			if ( registeredEventListener != null ) {
+				// eventType has already been registered;
+				return registeredEventListener; // EARLY RETURN
+			}
+			// eventType has not been registered yet.
+			// Its EventListenerGroupImpl will be created and added to registeredEventListeners below.
+		}
+		else {
+			// eventType is a custom EventType, and there is not enough space in
+			// registeredEventListeners to accommodate it.
+
+			// Allocate a new array to hold listener groups for *all* EventType values that currently exist.
+			// This way an existing, unregistered EventType with a larger ordinal will not require another
+			// allocation when it gets registered in the future.
+			final int sizeNew = Math.max( eventType.ordinal() + 1, EventType.values().size() );
```

Comment:
@Sanne, @NathanQingyangXu, before I address your questions, I want to make clear that making `EventType#addCustomEventType` `synchronized` and the line Sanne mentioned are related

When I removed `synchronized` from `EventType#addCustomEventType`, and ran the performance test, I found that there were cases where `eventType.ordinal() + 1` was greater than `EventType.values().size()`. As a result there were multiple `IndexOutOfBoundsException` thrown. 

That is why I changed:

`final int sizeNew = EventType.values().size();`

to: 

`final int sizeNew = Math.max( eventType.ordinal() + 1, EventType.values().size() );`

When I added `synchronized` back to `EventType#addCustomEventType`, that was no longer the case. I could have reverted back to what I had originally, but decided to keep it in case someone decided to removed `synchronized` from `EventType#addCustomEventType`.

Should I remove `synchronized` from `EventType#addCustomEventType`?

Reply:
@Sanne, I don't understand why your example would fail.

First off, each `EventType` should have at least one listener registered; otherwise, there's no reason to have an `EventType`. If no listener is registered for a particular `EventType`, then `EventListenerRegistryImpl#getEventListenerGroup` will throw `HibernateException`.

I opted to allocate space for all existing `EventType`s in `#registeredEventListeners` to reduce the number of times it needs to be re-allocated.

Obviously,  the number of allocations will depend on how the application creates `EventType`s and registers its listeners. If it creates all the custom `EventType`s first, then registers listeners, then only one re-allocation will be necessary.  If the application creates a custom `EventType`, then registers its listeners before creating the next custom `EventType`, then the array will need to be re-allocated for each custom `EventType`.

As I mentioned above, if `EventType#addCustomEventType` is not `synchronized`, then `eventType.ordinal() + 1` may be greater than `EventType.values().size()`. If it is synchronized, `eventType.ordinal() + 1` will always be <= `EventType.values().size()`.

AFAICT, the case that you mention should never happen. 

For 1), the array is allocated to accommodate an ordinal of 100.
For 2),  `sizeOriginal` is already 100, so the `if` condition is satisfied. `registeredEventListeners[ eventType.ordinal() ] == null`, so execution falls through to after the `else` clause (i.e., no re-allocation is done).
 
I'm not sure what you mean by "There's no need to literally "loop" among all of them, ...".

Could you add a failing test case to clarify?

Reply:
ok thanks Gail, I might have missed something. I'll merge it as is, if there's a problem we can follow up.